lib/iris/_data_manager.py

```diff
@@ -0,0 +1,392 @@
+# (C) British Crown Copyright 2010 - 2017, Met Office
+#
+# This file is part of Iris.
+#
+# Iris is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Iris is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Iris.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Management of common state and behaviour for cube and coordinate data.
+
+"""
+
+from __future__ import (absolute_import, division, print_function)
+from six.moves import (filter, input, map, range, zip)  # noqa
+import six
+
+import copy
+import numpy as np
+import numpy.ma as ma
+
+from iris._lazy_data import as_concrete_data, as_lazy_data, is_lazy_data
+
+
+# TODO: Complete the XXX doc-strings
+
+
+class DataManager(object):
+    """
+    XXX
+
+    """
+
+    def __init__(self, data, realised_dtype=None):
+        """
+        XXX
+
+        """
+        self._lazy_array = None
+        self._real_array = None
+        self.data = data
+
+        self._realised_dtype = None
+        self._dtype_setter(realised_dtype)
+
+        # Enforce the manager contract.
+        self._assert_axioms()
+
+    def __copy__(self):
+        """
+        Forbid :class:`DataManager` instance shallow-copy support.
+
+        """
+        name = type(self).__name__
+        emsg = ('Shallow-copy of {!r} is not permitted. Use '
+                'copy.deepcopy() or {}.copy() instead.')
+        raise copy.Error(emsg.format(name, name))
+
+    def __deepcopy__(self, memo):
+        """
+        Allow :class:`DataManager` instance deepcopy support.
+
+        Args:
+
+        * memo:
+            :class:`copy` memo dictionary.
+
+        """
+        return self._deepcopy(memo)
+
+    def __repr__(self):
+        fmt = '{cls}({self.core_data!r}{dtype})'
+        dtype = ''
+        if self._realised_dtype is not None:
+            dtype = ', realised_dtype={!r}'.format(self._realised_dtype)
+        result = fmt.format(self=self, cls=type(self).__name__, dtype=dtype)
+        return result
+
+    def _assert_axioms(self):
+        """
+        Definition of the manager state, that should never be violated.
+
```

Comment:
Personally, I think it's a bit much. It's a private method, so it's only there for the developer, and they can just read the code - so I'm seeing the doc-string giving some context to the dev not a literal translation of what it actually does.